### ISSUE TITLE
gallery.cnf is not verifying "sort" button

### ISSUE THREAD

**Issue by LukasReschke** [owncloud/gallery#445]
Data such as "type" is not checked properly, due to some hardening in core this does not lead directly to an exploitable XSS but should get addressed anyways.

![2015-10-19_16-36-49](https://cloud.githubusercontent.com/assets/878997/10580804/9e0e67be-767f-11e5-847a-f2cd2ae395c7.png)

Related code: https://github.com/owncloud/gallery/blob/d904a6a4af293e5730e397218c7e0713c3461215/js/galleryview.js#L322

Probably we should verify in the controller and also in the JS for data from the gallery config.

**Comment by LukasReschke** [comment]
Same is valid for `order` etc. - Best to just test a random config and append `">` to each values to see how it behaves :)

**Comment by oparoz** [comment]
> Probably we should verify in the controller and also in the JS for data from the gallery config.

Wouldn't just checking in the controller work? Any client would get valid data.

> Best to just test a random config and append "> to each values to see how it behaves :)

I'll check with the background colour, but unless styles can use Javascript, there should be no problem there.
I guess some regexp to make sure we're given alpha values doesn't hurt since it would also provide useful feedback to the user in case of failure.

**Comment by LukasReschke** [comment]
> Wouldn't just checking in the controller work? Any client would get valid data.

Sure. But I'm a fan of better twice than never, your choice :)

> I'll check with the background colour, but unless styles can use Javascript, there should be no problem there.

The colors are fine. `.css` is – generally speaking – not a problem. The problem is more the sorting stuff which you put into jQuery expressions which may get executed depending on which jQuery we in core do ship and which hardenings we do apply :)

**Comment by LukasReschke** [comment]
Well, except you consider horrible colors a problem :wink: 

![2015-10-19_15-59-45](https://cloud.githubusercontent.com/assets/878997/10581328/54724226-7682-11e5-9c9f-e2d74737e675.png)

**Comment by oparoz** [comment]
> Sure. But I'm a fan of better twice than never, your choice :)

I didn't want to duplicate the code, because basically, we're going to do the same thing in PHP and JS:
- Is the keyword in the list? Yes -> Add to config

but I guess an untrusted proxy could modify the code JSON in transit and it's best to check again at the other end...

So in the name of speed, I'm not going to use a sanitizer, but just check the values

### PHP
- [ ] Sort type `name || date`
- [ ] Sort order `asc || des`
- [ ] Colour `# + alpha`

### JS
- [ ] Sort type `name || date`
- [ ] Sort order `asc || des`
- [ ] Colour `# + alpha`

**Comment by LukasReschke** [comment]
:+1: - sounds good, evil proxy is out-of-scope so going with just PHP is okay. Well, or JS. Whatever works better for you :smile:

**Comment by oparoz** [comment]
> Well, or JS. Whatever works better for you :smile:

PHP, because one plan is to send a full gallery as an API response so that it can be embedded elsewhere, so it's best if we make an effort to send harmless content :)

**Comment by PVince81** [comment]
@oparoz any update ? not sure what the "ready" label means :smile: 
Just asking because this is set for 8.2.1 and we're already in RC1 :smile:

**Comment by oparoz** [comment]
Ready means this has been accepted for a sprint.
Since we're in RC1 already and this has a low impact, I'd prefer it if we delay the implementation of a fix to 8.2.2 since I should have more time then, but only if @LukasReschke is OK with it.

**Comment by ghost** [comment]
@LukasReschke are you ok with 8.2.2?

**Comment by LukasReschke** [comment]
Sure :smile:

**Comment by ghost** [comment]
@oparoz we're heading to the middle of the 8.2.2 timeline quickly, any thoughts on whether this will be part of it?

**Comment by oparoz** [comment]
@cmonteroluque Any idea when the RC will be packaged? I can work on this in a week and it should only take a couple of hours.

**Comment by ghost** [comment]
@oparoz by next Tuesday, 8 December

**Comment by oparoz** [comment]
Thanks!

**Comment by oparoz** [comment]
Blocked by https://github.com/owncloud/core/pull/20248 . We can't run integration tests **locally** any more.

**Comment by oparoz** [comment]
The PR https://github.com/owncloud/gallery/issues/477 fixes the issue and needs reviewing
